File: xkcd/src/main/java/xyz/jienan/xkcd/model/WhatIfModel.java
package xyz.jienan.xkcd.model;

import android.text.TextUtils;

import org.jsoup.nodes.Element;

import java.util.List;

import io.reactivex.Observable;
import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import io.reactivex.subjects.PublishSubject;
import xyz.jienan.xkcd.base.network.NetworkService;
import xyz.jienan.xkcd.base.network.WhatIfAPI;
import xyz.jienan.xkcd.model.persist.BoxManager;
import xyz.jienan.xkcd.model.util.WhatIfArticleUtil;

import static xyz.jienan.xkcd.base.network.NetworkService.XKCD_TOP;
import static xyz.jienan.xkcd.base.network.NetworkService.XKCD_TOP_SORT_BY_THUMB_UP;

public class WhatIfModel {

    private static WhatIfModel whatIfModel;

    private final WhatIfAPI whatIfAPI = NetworkService.getWhatIfAPI();

    private final BoxManager boxManager = BoxManager.getInstance();

    private final PublishSubject<WhatIfArticle> picsPipeline = PublishSubject.create();

    private WhatIfModel() {
        // no public constructor
    }

    public static WhatIfModel getInstance() {
        if (whatIfModel == null) {
            whatIfModel = new WhatIfModel();
        }
        return whatIfModel;
    }

    public void push(WhatIfArticle article) {
        picsPipeline.onNext(article);
    }

    public Observable<WhatIfArticle> observe() {
        return picsPipeline;
    }

    public Single<List<WhatIfArticle>> loadAllWhatIf() {
        return whatIfAPI.getArchive()
                .subscribeOn(Schedulers.io())
                .singleOrError()
                .map(WhatIfArticleUtil::getArticlesFromArchive)
                .map(boxManager::updateAndSaveWhatIf);
    }

    public Single<WhatIfArticle> loadLatest() {
        return loadAllWhatIf()
                .map(articleList -> articleList.get(articleList.size() - 1));
    }

    public Single<WhatIfArticle> loadArticle(long id) {
        return loadArticleContentFromDB(id)
                .switchIfEmpty(loadArticleFromAPI(id))
                .singleOrError()
                .observeOn(AndroidSchedulers.mainThread());
    }

    public List<WhatIfArticle> loadArticlesFromDB() {
        return boxManager.getWhatIfArchive();
    }

    public WhatIfArticle loadArticleFromDB(long id) {
        return boxManager.getWhatIf(id);
    }

    public Single<List<WhatIfArticle>> searchWhatIf(String query) {
        return Observable.just(boxManager.searchWhatIf(query))
                .singleOrError();
    }

    public Observable<WhatIfArticle> fav(long index, boolean isFav) {
        return Observable.just(boxManager.favWhatIf(index, isFav));
    }

    private Observable<WhatIfArticle> loadArticleFromAPI(long id) {
        return whatIfAPI.getArticle(id)
                .subscribeOn(Schedulers.io())
                .map(WhatIfArticleUtil::getArticleFromHtml)
                .map(Element::html)
                .map(content -> boxManager.updateAndSaveWhatIf(id, content));
    }

    private Observable<WhatIfArticle> loadArticleContentFromDB(long id) {
        WhatIfArticle article = boxManager.getWhatIf(id);
        return (article == null || TextUtils.isEmpty(article.content)) ? Observable.empty() : Observable.just(article);
    }

    public List<WhatIfArticle> getFavWhatIf() {
        return boxManager.getFavWhatIf();
    }

    public Observable<List<WhatIfArticle>> getThumbUpList() {
//        return xkcdApi.getTopXkcds(XKCD_TOP, XKCD_TOP_SORT_BY_THUMB_UP)
//                .subscribeOn(Schedulers.io())
//                .map(boxManager::updateAndSave);
        return Observable.empty();
    }
}

File: xkcd/src/main/java/xyz/jienan/xkcd/model/WhatIfModel.kt
package xyz.jienan.xkcd.model

import android.text.TextUtils
import io.reactivex.Completable
import io.reactivex.Flowable
import io.reactivex.Maybe
import io.reactivex.Observable
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import io.reactivex.subjects.PublishSubject
import xyz.jienan.xkcd.base.network.NetworkService
import xyz.jienan.xkcd.base.network.WHAT_IF_THUMBS_UP
import xyz.jienan.xkcd.base.network.WHAT_IF_TOP
import xyz.jienan.xkcd.base.network.XKCD_TOP_SORT_BY_THUMB_UP
import xyz.jienan.xkcd.model.persist.BoxManager
import xyz.jienan.xkcd.model.util.WhatIfArticleUtil


object WhatIfModel {

    private val whatIfAPI = NetworkService.whatIfAPI

    private val picsPipeline = PublishSubject.create<WhatIfArticle>()

    private val zoomPipeline = PublishSubject.create<Int>()

    val favWhatIf: List<WhatIfArticle>
        get() = BoxManager.favWhatIf

    val thumbUpList: Single<List<WhatIfArticle>>
        get() = whatIfAPI.getTopWhatIfs(WHAT_IF_TOP, XKCD_TOP_SORT_BY_THUMB_UP)
                .subscribeOn(Schedulers.io())

    val untouchedList: List<WhatIfArticle>
        get() = BoxManager.untouchedArticleList

    fun push(article: WhatIfArticle) {
        picsPipeline.onNext(article)
    }

    fun setZoom(zoom: Int) {
        zoomPipeline.onNext(zoom)
    }

    fun observeZoom(): Observable<Int> {
        return zoomPipeline
    }

    fun observe(): Observable<WhatIfArticle> {
        return picsPipeline
    }

    fun loadAllWhatIf(): Single<List<WhatIfArticle>> {
        return whatIfAPI.archive
                .subscribeOn(Schedulers.io())
                .map { WhatIfArticleUtil.getArticlesFromArchive(it) }
                .map { BoxManager.updateAndSaveWhatIf(it.toMutableList()) }
    }

    fun loadLatest(): Single<WhatIfArticle> {
        return loadAllWhatIf()
                .map { it.last() }
    }

    fun loadArticle(id: Long): Single<WhatIfArticle> {
        return loadArticleContentFromDB(id)
                .switchIfEmpty(loadArticleFromAPI(id))
                .observeOn(AndroidSchedulers.mainThread())
    }

    fun loadArticlesFromDB(): List<WhatIfArticle>? {
        return BoxManager.whatIfArchive
    }

    fun loadArticleFromDB(id: Long): WhatIfArticle? {
        return BoxManager.getWhatIf(id)
    }

    fun searchWhatIf(query: String, option: String): Single<MutableList<WhatIfArticle>> {
        return Observable.just(BoxManager.searchWhatIf(query, option))
                .singleOrError()
    }

    fun fav(index: Long, isFav: Boolean): Observable<WhatIfArticle> {
        return Observable.just(BoxManager.favWhatIf(index, isFav)!!)
    }

    /**
     * @param index
     * @return thumb up count
     */
    fun thumbsUp(index: Long): Single<Long> {
        return whatIfAPI.thumbsUpWhatIf(WHAT_IF_THUMBS_UP, index.toInt())
                .subscribeOn(Schedulers.io())
                .doOnSubscribe { BoxManager.likeWhatIf(index) }
                .map { it.thumbCount }
    }

    fun fastLoadWhatIfs(index: Long): Completable =
            Flowable.just(index)
                    .flatMap {
                        if (index == 0L) {
                            loadLatest().map { it.num }
                                    .toFlowable()
                                    .flatMap { Flowable.rangeLong(1, it) }
                        } else {
                            Flowable.rangeLong(1, index)
                        }
                    }.flatMapSingle {
                        if (loadArticleFromDB(it) == null || loadArticleFromDB(it)!!.content.isNullOrBlank()) {
                            loadArticleFromAPI(it)
                                .map { index }
                                .onErrorReturnItem(index)
                        } else {
                            Single.just(it)
                        }
                    }
                    .toList()
                    .ignoreElement()
                    .onErrorComplete()

    private fun loadArticleFromAPI(id: Long): Single<WhatIfArticle> {
        return whatIfAPI.getArticle(id)
                .subscribeOn(Schedulers.io())
                .map { WhatIfArticleUtil.getArticleFromHtml(it) }
                .map { it.html() }
                .map { BoxManager.updateAndSaveWhatIf(id, it) }
    }

    private fun loadArticleContentFromDB(id: Long): Maybe<WhatIfArticle> {
        val article = BoxManager.getWhatIf(id)
        return if (article == null || TextUtils.isEmpty(article.content)) Maybe.empty() else Maybe.just(article)
    }
}
